File: src/data/codes/logistic-regression.m
clc;
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'diabetes_data.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(6);
    n = 160;

    Age = randi([20 70], n, 1);
    BMI = 18 + 18*rand(n,1);
    Glucose = randi([75 200], n, 1);
    Blood_Pressure = randi([60 100], n, 1);

    % Create probability using logistic-type relation
    z = -18 + 0.05*Age + 0.18*BMI + 0.04*Glucose + 0.03*Blood_Pressure;
    p = 1 ./ (1 + exp(-z));
    Diabetes = double(rand(n,1) < p);

    data = table(Age, BMI, Glucose, Blood_Pressure, Diabetes);
else
    data = readtable(csvFile);
    % Expected columns:
    % Age, BMI, Glucose, Blood_Pressure, Diabetes
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.Age, data.BMI, data.Glucose, data.Blood_Pressure];
Y = data.Diabetes;

% Normalize features
X = normalize(X);

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = fitclinear(X_train, Y_train, ...
    'Learner', 'logistic', ...
    'Regularization', 'ridge');

%% =========================
% PREDICTION
% ==========================
[label, score] = predict(model, X_test);
Y_pred = label;              % 0 or 1
probDiabetes = score(:,2);   % probability for class 1

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

TP = confMat(2,2);
TN = confMat(1,1);
FP = confMat(1,2);
FN = confMat(2,1);

precision = TP / (TP + FP + eps);
recall    = TP / (TP + FN + eps);
f1        = 2 * precision * recall / (precision + recall + eps);

disp('Confusion Matrix:');
disp(confMat);

fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

%% =========================
% VISUALIZATION
% ==========================
figure;
confusionchart(Y_test, Y_pred);
title('Logistic Regression Confusion Matrix');

figure;
scatter(1:length(Y_test), probDiabetes, 'filled');
xlabel('Test Sample');
ylabel('Predicted Probability of Diabetes');
title('Predicted Probability');
grid on;

File: src/data/codes/logistic-regression.py
"""Logistic regression (L2 / ridge)
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(6)

if USE_SYNTHETIC:
    n = 160
    age = RNG.integers(20, 71, size=n)
    bmi = RNG.uniform(18, 36, size=n)
    glucose = RNG.integers(75, 201, size=n)
    bp = RNG.integers(60, 101, size=n)
    z = -18 + 0.05 * age + 0.18 * bmi + 0.04 * glucose + 0.03 * bp
    p = 1 / (1 + np.exp(-z))
    diabetes = (RNG.random(n) < p).astype(int)
    data = pd.DataFrame(
        {"Age": age, "BMI": bmi, "Glucose": glucose, "Blood_Pressure": bp, "Diabetes": diabetes}
    )
else:
    data = pd.read_csv("diabetes_data.csv")

print(data.head())
data = data.dropna().drop_duplicates()

X = data[["Age", "BMI", "Glucose", "Blood_Pressure"]].values
y = data["Diabetes"].values
scaler = StandardScaler()
X = scaler.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = LogisticRegression(penalty="l2", solver="lbfgs", max_iter=1000, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
proba = model.predict_proba(X_test)[:, 1]

cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", cm)
print(f"\nAccuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("Logistic Regression Confusion Matrix")

fig2, ax2 = plt.subplots()
ax2.scatter(np.arange(len(y_test)), proba, alpha=0.7)
ax2.set_xlabel("Test Sample")
ax2.set_ylabel("Predicted Probability of Diabetes")
ax2.set_title("Predicted Probability")
ax2.grid(True)
plt.show()
